presto-main/src/main/java/com/facebook/presto/sql/planner/RelationPlanner.java

```diff
@@ -725,14 +725,18 @@ private RelationPlan processAndCoerceIfNecessary(Relation node, Void context)
 
     private RelationPlan addCoercions(RelationPlan plan, Type[] targetColumnTypes)
     {
-        List<VariableReferenceExpression> oldVariables = plan.getFieldMappings();
-        RelationType oldDescriptor = plan.getDescriptor().withOnlyVisibleFields();
-        verify(targetColumnTypes.length == oldVariables.size());
+        RelationType oldRelation = plan.getDescriptor();
+        List<VariableReferenceExpression> oldVisibleVariables = oldRelation.getVisibleFields().stream()
+                .map(oldRelation::indexOf)
+                .map(plan.getFieldMappings()::get)
+                .collect(toImmutableList());
+        RelationType oldRelationWithVisibleFields = plan.getDescriptor().withOnlyVisibleFields();
+        verify(targetColumnTypes.length == oldVisibleVariables.size());
```

Comment:
I'm not confident about this change.  RelationPlanner.visitTable() explicitly calls addCoercions including the hidden fields in the type.  This change seems like it would break the case were there are hidden columns.
https://github.com/prestodb/presto/blob/3c2d9d80e4aba2c7fcba2a8b1df28c4c66875221/presto-main/src/main/java/com/facebook/presto/sql/planner/RelationPlanner.java#L154-L155

Reply:
If you refer to a hidden field explicitly in the subquery, the field will no longer be hidden. For example, if you do 
```
WITH t(x, y, z) as (SELECT x, y, "some_hidden_field" FROM t)
```
The `"some_hidden_field"` is no longer hidden in the `scope.getRelationType()`. The only situation that the named query could include hidden field is when the output is not explicit. Looking at the syntax, I think this would only happen if the named query is a table reference. For table reference, not that in `StatementAnalyzer` https://github.com/prestodb/presto/blob/3c2d9d80e4aba2c7fcba2a8b1df28c4c66875221/presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java#L2106

We actually check against visible field count. So considering all fields with guarantee failure if there's hidden field.

Reply:
The explanation makes sense.  Is it possible to fix this in the AST builder, then so we don't even have the hidden column for with clause output? It seems like that would be a better solution.  Otherwise, this looks good.

Reply:
the `visitTable` in `AstBuilder` doesn't really retrieve metadata from table. Unless we change that this would be hard to do. I'll merge this PR as is. Thanks!